torch/csrc/distributed/c10d/CUDASymmetricMemoryOps.cu

```diff
@@ -693,6 +695,123 @@ at::Tensor stream_write_value32_(
   return input;
 }
 
+std::tuple<at::Tensor, at::Tensor> _all_gather_mm(
+    at::Tensor a_shard,
+    at::Tensor b,
+    std::string group_name) {
+  c10::cuda::CUDAGuard guard(a_shard.device());
+
+  auto symm_mem = c10d::symmetric_memory::rendezvous(a_shard, group_name);
+  TORCH_CHECK(
+      symm_mem != nullptr,
+      "_all_gather_mm: input must be allocated with empty_strided_p2p().");
+
+  int rank = symm_mem->get_rank();
+  int world_size = symm_mem->get_world_size();
+
+  auto current_stream = at::cuda::getCurrentCUDAStream();
+
+  static std::optional<at::cuda::CUDAStream> copy_stream = std::nullopt;
+  static std::optional<at::cuda::CUDAStream> signal_stream = std::nullopt;
+  if (!copy_stream.has_value()) {
+    copy_stream = c10::cuda::getStreamFromPool(-1, a_shard.device().index());
+    signal_stream = c10::cuda::getStreamFromPool(-1, a_shard.device().index());
+  }
+
+  symm_mem->barrier(0, 0);
+  {
+    auto event = at::cuda::CUDAEvent();
+    event.record(current_stream);
+    event.block(*copy_stream);
+  }
```

Comment:
There is a `CUDAStream::waitStream` that would make this more readable
https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/torch/csrc/jit/cuda/cuda.h#L132-L137